### PR TITLE
Fix for Duplicate Webhook deliveries

### LINKED ISSUE #24729: Product.updated webhook triggers twice
**Describe the bug**
When I use the webhook product.updated it triggers twice. 

**To Reproduce**
1. Create a webhook for product.updated
2. Update a product's name
3. Watch for the error

**screenshots**
Schermafbeelding 2019-10-01 om 10 09 33

Schermafbeelding 2019-10-01 om 10 09 41

**Expected behavior**
I expected that the webhook would trigger twice

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I searched for the same issue in this repo, but all issues were old and closed.
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
```
`
### WordPress Environment ###

WordPress address (URL): https://****.com
Site address (URL): https://****.com
WC Version: 3.7.0
REST API Version: ✔ 1.0.2
Log Directory Writable: ✔
WP Version: 5.2.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.15.0
PHP Version: 7.1.6-1~ubuntu16.04.1+deb.sury.org+1 - We recommend using PHP version 7.2 or above for greater performance and security. How to update your PHP version
PHP Post Max Size: 64 MB
PHP Time Limit: 300
PHP Max Input Vars: 1000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.1.35-MariaDB-1~xenial
Max Upload Size: 64 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.7.0
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 4.63MB
Database Data Size: 3.13MB
Database Index Size: 1.50MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.14MB + Index: 0.19MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.03MB + Index: 0.08MB + Engine InnoDB
wp_postmeta: Data: 0.14MB + Index: 0.11MB + Engine InnoDB
wp_posts: Data: 0.16MB + Index: 0.13MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (6) ###

Jetpack by WordPress.com: by Automattic – 7.7.1
Mailtrap for WordPress: by Eduardo Marcolino – 0.3
WooCommerce Admin: by WooCommerce – 0.18.0
WooCommerce PayPal Checkout Gateway: by WooCommerce – 1.6.17
WooCommerce Services: by Automattic – 1.21.1
WooCommerce: by Automattic – 3.7.0

### Inactive Plugins (0) ###

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 2.5.3
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 267
Oldest: 2019-09-20 12:46:03 +0000
Newest: 2019-10-01 07:54:14 +0000

Pending: 2
Oldest: 2019-10-01 08:16:38 +0000
Newest: 2019-10-01 08:16:38 +0000

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 0
Oldest: –
Newest: –

`
```
</details>

### PR DESCRIPTION
**Pull request by thejamescollins**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Currently, WooCommerce may trigger multiple webhook deliveries for a single webhook if the webhook topic has multiple hooks defined in [WC_Webhook::get_topic_hooks()](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-webhook.php#L923), for example:

- Coupon Created (2)
- Coupon Updated (2)
- Customer Created (3)
- Customer Updated (2)
- Order Updated (2)
- Product Created (3)
- Product Updated (3)

If using the default Action Scheduler (async) delivery mechanism, this is less likely to be an issue due to the changes introduced in #22087, however, if someone is using a different mechanism such as synchronous (immediate) delivery then this is definitely a major issue.

Even if using the default Action Scheduler based delivery mechanism there is a chance that duplicate deliveries could occur if the first delivery is queued, then the scheduled action is processed, and then the duplicate delivery is queued.

This PR contains changes so that the `woocommerce_webhook_process_delivery` hook should only trigger once per webhook and webhook arg, which improves delivery reliability regardless of whether the user is using Action Scheduler, immediate (synchronous) delivery or their own Queueing/Delivery mechanism.

cc @thenbrent (who authored that previous change in #22087 in case he has feedback here).

Closes #24729 

### How to test the changes in this Pull Request:

The new unit test added verifies that the bug exists and that this change fixes it.

However in order to test things manually:

1.  add this line of code to instruct your WooCommerce installation to delivery webhooks synchronously:

```
add_filter( 'woocommerce_webhook_deliver_async', '__return_false' );
```

2. Create a new Webhook, choosing a webhook topic that has multiple hooks defined. For example, Customer Created (which has 3 defined).

3. Manually create a customer in your WooCommerce Store, and check the webhook delivery logs to verify that the customer was delivered once (not 3 times).

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Duplicate webhook deliveries

### Next Steps

If this PR is merged,  I propose that the code introduced in #22087 is no longer necessary and could be reverted.

### CODE REVIEW

File: includes/class-wc-webhook.php [hunk 1]
```diff
@@ -103,6 +111,9 @@ public function process( $arg ) {
 			return;
 		}
 
+		// Mark this $arg as processed to ensure it doesn't get processed again within the current request.
+		$this->processed[] = $arg;
```
Comment (peterfabian) [inline includes/class-wc-webhook.php]: Why not add the `$arg` as a key and then test via `isset`/`array_key_exists`? I guess that should be quite a bit faster, potentially. Maybe unnecessary optimization...

**Review by peterfabian — COMMENTED** [review]
Thanks for your contribution, it fixes the problem as far as I can see. One small note below.

File: includes/class-wc-webhook.php (hunk 1 above)
Comment (peterfabian) [reply includes/class-wc-webhook.php]: To answer my own question, array keys can't be e.g. arrays or objects and $arg is of mixed type, so this way it's safer, especially for hooks tied to actions which can have objects and arrays as arguments. 
The number of hooks running within one request probably won't be large anyway.

**Review by peterfabian — APPROVED** [review]
Looks good, thanks.

**Comment by om4james** [comment]
Thank you @peterfabian :)

**Comment by vedanshujain** [comment]
@om4james @peterfabian I tested multiple scenarios, and I think we should keep throttling changes introduced in #22087. This is because, while this fix handles cases where duplicate webhooks are getting scheduled from the same request, throttling code handles one additional edge case where duplicate async webhooks are getting scheduled from different requests. Also, since the fix introduced in this PR is executed before the throttling fix, it will not be executed at all in most cases so there aren't any performance implications as well.

Right now if two request schedule sync webhooks, they will both be delivered, but I think that's outside of the scope of this PR and we can address this later if its a common enough scenario.

In short, let's keep both the fixes because the older code handles an additional edge case.
